content/blog/2023-12-11-bloke/index.md

```diff
@@ -0,0 +1,372 @@
+
+## Correctness Cost
+
+Calculating correctness of a program is inefficient if not hard. This is why the STOKE correctness cost function has a fast but unsound validation component using test cases and a slow but sound verification component using satisfiability modulo theory (SMT) solvers:
+
+$$
+\mathrm{eq}(x)= \begin{cases}\text{val}(x) & \text { if } \text{val}(x)>0; \\ \text{ver}(x) & \text { otherwise }\end{cases}
+$$
+
+Given testcases $\tau=(i\_{1}, o\_{1}), \ldots,(i\_{n}, o\_{n})$ where $o\_{j}$ is the output of the initial program $x\_{0}$ given input $i\_{j}$, the validation component val: $X \rightarrow \mathbf{R}^{\geqslant 0}$ is defined as the sum over the Hamming distances $d$ of the outputs, including additional cost if the program produces an error:
+
+$$
+\text{val}(x)=\sum\_{(i, o) \in \tau} d(x(i), o)+\text{err}(x, i)
+$$
+
+The error component err : $X \times I \rightarrow \mathbf{R}^{\geqslant 0}$, with input space $I$, takes in a program $x \in X$ and an input $i \in I$, and computes an error term $e \in \mathbf{R}^{\geqslant 0}$ such that $\text{err}(x, i)=0$ if and only if $x$ does not produce an error on input $i$.
+
+The verification function of STOKE calculates whether program $x$ is equivalent to the initial program $x\_{0}$ :
+
+$$
+\text{ver}(x)= \begin{cases}1 & \text { if } \exists i \in I[x(i) \neq x\_{0}(i)]; \\ 0 & \text { otherwise }\end{cases}
+$$
+
+STOKE uses the STP theorem prover [2] to calculate the verification function. Code sequences are translated to SMT formulae and is queried whether there is an input $i \in I$ such that $x$ and $x\_{0}$ produce different outputs or side effects.
+
+If the theorem prover returns sat, then there exists a counterexample $c \in I$ such that $x(c) \neq x\_{0}(x)$, and can then be added to the test cases used in the faster val function.
+
+If the prover returns unsat, then that means for all inputs $i \in I, x(i)=x\_{0}(i)$, and therefore the two programs are equivalent.
+
+## Performance Cost
+
+Instead of expensive JIT compilation, STOKE opts a simple heuristic to statically approximate program runtime:
+
+$$
+\text{perf}(x)=\sum\_{i \in \text{instr}(x)} \text { latency }(i)
+$$
+
+where latency $(i)$ is the average latency of an instruction $i$.
+
+## Two-Stage Process
+
+STOKE uses a two-stage process to find optimized programs. STOKE first has the synthesis phase that focus solely on correctness. STOKE then uses the optimization phase to introduce the performance component of the cost function, which allows the search for the fastest program in the correct program space.
+
+In short, the synthesis and the optimization phases have these cost functions:
+
+- synthesis, eq $(x)$; and
+- optimization, eq $(x)+\text{perf}(x)$.
```

Comment:
I'm not quite sure how to parse these equations. Is "eq" supposed to be inside the math delimiters? Like $\text{eq}(x) + \text{perf}(x)$, for example? (I realize this might look similar when rendered.)

Reply:
Yes, the interpretation is correct. I have included `eq` within the delimeters